### PR TITLE
Adding matchers for assertions based on enzyme selectors

### LINKED ISSUE #260: Add support for containsMatchingElement
A common use case with Enzyme is to test that a wrapper contains an element that matches an enzyme selector. This can be done manually, using [`containsMatchingElement`](https://airbnb.io/enzyme/docs/api/ShallowWrapper/containsMatchingElement.html), but this does not produce very readable errors.

A matcher should be created that replicates the `containsMatchingElement` behavior.

```js
const container = shallow(<MyContainer />);

expect(container.containsMatchingElement('.item').toBe(true);
```

This produces the error, "Expected false to be true."

Ideally, you'd be able to do this with a matcher.
```js
const container = shallow(<MyContainer />);

expect(container).toContainMatchingElement('.item');
```

This could produce an error like, "Expected <MyContainer /> to contain an element matching '.item'."

Additionally, a similar matcher should also be created to assert `n` matches.

Possible names for the new matchers:
- `toContainMatchingElement(s)`
- `toContainMatchingReactElement(s)` (inline with `toContainReact`)
- `toContainMatch(es)`

### Implementation notes

For the negative match, the `toContainMatchingElement` matcher will either have to use `.find` and validate the length of the result or handle the error produced by `containsMatchingElement` when more than one match is found.

### PR DESCRIPTION
**Pull request by GreenGremlin**
Adding three new matchers.

### toContainMatch('.foo')

Asserts that the given wrapper contains at least one match for the given selector.

### toContainMatches(2, '.foo')

Asserts that the given wrapper contains a given number of matches for the given selector.

### toContainExactlyOneMatch('.foo')

Asserts that the given wrapper contains exactly one match for the given selector.

Resolves #260

### CODE REVIEW

File: packages/enzyme-matchers/src/assertions/toContainExactlyOneMatch.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+import type { EnzymeObject, Matcher } from '../types';
+import toContainMatches from './toContainMatches';
+
+function toContainExactlyOneMatch(
+  enzymeWrapper: EnzymeObject,
+  selector: String
+): Matcher {
+  return toContainMatches(enzymeWrapper, 1, selector);
```
Comment (blainekasten) [inline packages/enzyme-matchers/src/assertions/toContainExactlyOneMatch.js]: clever, nice.

File: packages/enzyme-matchers/src/assertions/toContainMatches.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import type { EnzymeObject, Matcher } from '../types';
+import html from '../utils/html';
+import getNodeName from '../utils/name';
+import single from '../utils/single';
+
+function toContainMatches(
+  enzymeWrapper: EnzymeObject,
+  n: Integer,
```
Comment (blainekasten) [inline packages/enzyme-matchers/src/assertions/toContainMatches.js]: is Integer a valid primitive for flow? I thought it was just `number`. Can you run `yarn flow` and see if there are errors?

File: packages/enzyme-matchers/src/index.js [hunk 3]
```diff
@@ -30,6 +33,9 @@ const assertions = {
   toBeChecked,
   toBeDisabled,
   toBeEmptyRender,
+  toContainMatch,
+  toContainMatches,
+  toContainExactlyOneMatch,
```
Comment (blainekasten) [inline packages/enzyme-matchers/src/index.js]: we need to add these matchers to the typescript spec here: https://github.com/FormidableLabs/enzyme-matchers/blob/master/packages/jest-enzyme/src/index.d.ts

File: packages/enzyme-matchers/src/assertions/toContainMatches.js (hunk 2 above)
Comment (GreenGremlin) [reply packages/enzyme-matchers/src/assertions/toContainMatches.js]: Ah, it is not. Also, String should be string.

Fixed

File: packages/enzyme-matchers/src/index.js (hunk 3 above)
Comment (GreenGremlin) [reply packages/enzyme-matchers/src/index.js]: Done

**Comment by GreenGremlin** [comment]
> I'm wondering if we should make it a little more descriptive toContainElementThatMatches toContainElementsThatMatch toContainExactlyOneElementThatMatches, not sure if that verbosity is beneficial or hurtful...

I absolutely agree. I was thinking the same thing, but never circled back to make up my mind. I went with `toContainMatchingElement`, `toContainMatchingElements`, and `toContainExactlyOneMatchingElement`; to more closely match the enzyme api.

**Comment by blainekasten** [comment]
Ooo, I like that! Awesome work. Let's roll with this! I'll release a new minor now

**Pull request merged**